### PR TITLE
Password mail with IP addresses.

### PR DESCRIPTION
**Pull request by begrif**
Save the user's IP address and browser `User-Agent:` when creating password reset tokens. Then when sending password reset email, include those in the message sent.

### CODE REVIEW

File: db/migrate/20170816343911_add_password_change_ip_to_email_tokens.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class AddPasswordChangeIpToEmailTokens < ActiveRecord::Migration
+  def up
+    add_column :email_tokens, :remote_ip, :inet
```
Comment (tgxworld) [inline db/migrate/20170816343911_add_password_change_ip_to_email_tokens.rb]: Can the following columns be `null`? :thought_balloon:

File: app/controllers/session_controller.rb [hunk 2]
```diff
@@ -248,7 +248,8 @@ def forgot_password
     user = User.find_by_username_or_email(params[:login])
     user_presence = user.present? && user.id > 0 && !user.staged
     if user_presence
-      email_token = user.email_tokens.create(email: user.email)
+      user_agent = request.user_agent || '(unidentified browser)'
```
Comment (tgxworld) [inline app/controllers/session_controller.rb]: `(unidentified browser)` should probably be translated using `I18n` here

File: db/migrate/20170816343911_add_password_change_ip_to_email_tokens.rb (hunk 1 above)
Comment (begrif) [inline db/migrate/20170816343911_add_password_change_ip_to_email_tokens.rb]: Yes.

File: app/controllers/session_controller.rb (hunk 2 above)
Comment (begrif) [inline app/controllers/session_controller.rb]: Sure, I can do that.

**Comment by begrif** [comment]
I think the ideal situation is that this is logged and admins of more technical forums can enable it in emails at their sites, but it is not shown by default.

**Comment by SamSaffron** [comment]
we could possibly enable this out of the box for moderators and admins. I am not sure. I think you should split this up. 

The first non-controversial change is just to log somewhere. cause then you could look at the user log and see that someone is trying to compromise an account and see where it is.

**Comment by coding-horror** [comment]
Sure I support displaying the extra info for staff (mods/admins) only at
first. That is a fine idea.

**Comment by pain0x0** [comment]
yes long user agent strings can confuse non-tech people but still would be ok to still have it i guess?
and yes a panel in admin section where it logs password reset and show's ip etc would be good and having option where user can report if its not them

**Comment by coding-horror** [comment]
As long as user agent is only shown to staff password resets I support merging this in @ZogStriP

File: app/mailers/user_notifications.rb [hunk 3]
```diff
@@ -48,10 +48,25 @@ def confirm_new_email(user, opts = {})
   end
 
   def forgot_password(user, opts = {})
+    remote_ip = I18n.t('no_remote_ip')
+    user_agent = I18n.t('no_browser_user_agent')
+
+    if opts[:email_token] then
```
Comment (ZogStriP) [inline app/mailers/user_notifications.rb]: We only use `then` in inline `case` statements.

**Comment by ZogStriP** [comment]
Also, this needs to be only shown to `staff` members.

**Comment by SamSaffron** [comment]
We can't keep this open forever, closing for now, @begrif can you resubmit a PR when it is ready.

**Pull request closed without merging**